### ISSUE TITLE
Auto completion of `async fn xxxx` for desugared `async fn in trait` form

### ISSUE THREAD

**Issue by hyf0** [rust-lang/rust-analyzer#17719]
Hi. I'm thinking about improving the dx of writing `impl DesugaredAsyncTrait for Foo`. So far the rust analyzer will only infer auto completion like

![image](https://github.com/user-attachments/assets/525418e3-8561-4921-97ca-c7e71dacf228)

However, I found that the compiler will allow the following code compiles. ~If this is an expected behavior~(It's expected, see the following zulipchat link), maybe it's ok to make rust analyzer to support auto completion that starts with `async fn foo`( the sugar version rather than only desugared version like now) for `DesugaredAsyncTrait `.

This will surely improve the dx of writing `impl DesugaredAsyncTrait for XXX` greatly without the need to write `impl Future<Output=XXXX> + Send` over and over again(Though there are also auto completed by the rust-analyzer :).

```rust
use std::future::Future;

trait DesugaredAsyncTrait {
    fn foo(&self) -> impl Future<Output = usize> + Send;
    fn bar(&self) -> impl Future<Output = usize> + Send;
}

struct Foo;

impl DesugaredAsyncTrait for Foo {
    fn foo(&self) -> impl Future<Output = usize> + Send {
        async { 1 }
    }

    //
    async fn bar(&self) -> usize {
        1
    }
}

fn main() {
    let fut = Foo.bar();
    fn _assert_send<T: Send>(_: T) {}
    _assert_send(fut);
}
```

Related:
- https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Auto.20infered.20.60Send.60.20in.20.60impl.20DesugaredAsync.60.2E.20Expected.3F
- https://x.com/Gankra_/status/1141409682017308672

**Comment by hyf0** [comment]
There's a question on `Quick Fix "Implement missing members"` if we support this PR.

On "Implement missing members", should we fill them in sugar version or original form for `async fn in trait` in desugar form?

**Comment by Veykril** [comment]
I'd say sugared version, and then if we don't have it yet, have a desugar assist that goes from `async` to `impl Future` (I think we already have this though)

**Comment by lnicola** [comment]
Fixed in #17737
